### PR TITLE
refactor(core): make the error messages tree shakable

### PR DESCRIPTION
**Pull request by ramthir**
Long error messages can be tree-shaken in the production build and replaced with error codes.

See: https://github.com/angular/angular/pull/44219#issuecomment-983216374

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?
Long error messages are not tree-shakable in the production build.

Issue Number: N/A

## What is the new behavior?
Long error messages are tree-shaken in the production build and replaced with error codes.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/core/src/render3/error_code.ts [hunk 1]
```diff
@@ -42,6 +47,16 @@ export const enum RuntimeErrorCode {
   // i18n Errors
 
   // Compilation Errors
+
+  // Other
+  INVALID_ARGUMENTS = '900',
```
Comment (ramthir) [inline packages/core/src/render3/error_code.ts:52]: @AndrewKushnir I need some help with the error codes, not sure about the sections of these and if we can group some of them.

Comment (AndrewKushnir) [reply packages/core/src/render3/error_code.ts:52]: @ramthir thanks for the comment, I can provide some feedback during the review process.

File: packages/core/src/change_detection/differs/default_iterable_differ.ts [hunk 2]
```diff
@@ -152,8 +153,10 @@ export class DefaultIterableDiffer<V> implements IterableDiffer<V>, IterableChan
   diff(collection: NgIterable<V>|null|undefined): DefaultIterableDiffer<V>|null {
     if (collection == null) collection = [];
     if (!isListLikeIterable(collection)) {
-      throw new Error(
-          `Error trying to diff '${stringify(collection)}'. Only arrays and iterables are allowed`);
+      const errorMessage = (typeof ngDevMode === 'undefined' || ngDevMode) ?
+          `Error trying to diff '${stringify(collection)}'. Only arrays and iterables are allowed` :
+          '';
+      throw new RuntimeError(RuntimeErrorCode.INVALID_ARGUMENTS, errorMessage);
```
Comment (AndrewKushnir) [inline packages/core/src/change_detection/differs/default_iterable_differ.ts:159]:
```suggestion
      throw new RuntimeError(RuntimeErrorCode.INVALID_DIFFER_INPUT, errorMessage);
```

File: packages/core/src/change_detection/differs/default_keyvalue_differ.ts [hunk 3]
```diff
@@ -79,8 +81,10 @@ export class DefaultKeyValueDiffer<K, V> implements KeyValueDiffer<K, V>, KeyVal
     if (!map) {
       map = new Map();
     } else if (!(map instanceof Map || isJsObject(map))) {
-      throw new Error(
-          `Error trying to diff '${stringify(map)}'. Only maps and objects are allowed`);
+      const errorMessage = (typeof ngDevMode === 'undefined' || ngDevMode) ?
+          `Error trying to diff '${stringify(map)}'. Only maps and objects are allowed` :
+          '';
+      throw new RuntimeError(RuntimeErrorCode.INVALID_ARGUMENTS, errorMessage);
```
Comment (AndrewKushnir) [inline packages/core/src/change_detection/differs/default_keyvalue_differ.ts:87]:
```suggestion
      throw new RuntimeError(RuntimeErrorCode.INVALID_DIFFER_INPUT, errorMessage);
```

File: packages/core/src/change_detection/differs/iterable_differs.ts [hunk 4]
```diff
@@ -250,8 +251,11 @@ export class IterableDiffers {
     if (factory != null) {
       return factory;
     } else {
-      throw new Error(`Cannot find a differ supporting object '${iterable}' of type '${
-          getTypeNameForDebugging(iterable)}'`);
+      const errorMessage = (typeof ngDevMode === 'undefined' || ngDevMode) ?
+          `Cannot find a differ supporting object '${iterable}' of type '${
+              getTypeNameForDebugging(iterable)}'` :
+          '';
+      throw new RuntimeError(RuntimeErrorCode.NO_SUPPORTING_FACTORY, errorMessage);
```
Comment (AndrewKushnir) [inline packages/core/src/change_detection/differs/iterable_differs.ts:258]:
```suggestion
      throw new RuntimeError(RuntimeErrorCode.NO_SUPPORTING_DIFFER_FACTORY, errorMessage);
```

File: packages/core/src/change_detection/differs/keyvalue_differs.ts [hunk 5]
```diff
@@ -181,6 +183,9 @@ export class KeyValueDiffers {
     if (factory) {
       return factory;
     }
-    throw new Error(`Cannot find a differ supporting object '${kv}'`);
+    const errorMessage = (typeof ngDevMode === 'undefined' || ngDevMode) ?
+        `Cannot find a differ supporting object '${kv}'` :
+        '';
+    throw new RuntimeError(RuntimeErrorCode.NO_SUPPORTING_FACTORY, errorMessage);
```
Comment (AndrewKushnir) [inline packages/core/src/change_detection/differs/keyvalue_differs.ts:189]:
```suggestion
    throw new RuntimeError(RuntimeErrorCode.NO_SUPPORTING_DIFFER_FACTORY, errorMessage);
```

File: packages/core/src/di/injector_compatibility.ts [hunk 6]
```diff
@@ -98,15 +102,15 @@ export function ɵɵinject<T>(token: ProviderToken<T>, flags = InjectFlags.Defau
  * @codeGenApi
  */
 export function ɵɵinvalidFactoryDep(index: number): never {
-  const msg = ngDevMode ?
+  const errorMessage = (typeof ngDevMode === 'undefined' || ngDevMode) ?
```
Comment (AndrewKushnir) [inline packages/core/src/di/injector_compatibility.ts:105]: Since this is a part of the generated code, I think we should avoid using extra classes here. Let's revert to the previous state.

File: packages/core/src/di/injector_compatibility.ts [hunk 7]
```diff
@@ -58,7 +59,10 @@ export function injectInjectorOnly<T>(token: ProviderToken<T>, flags?: InjectFla
 export function injectInjectorOnly<T>(token: ProviderToken<T>, flags = InjectFlags.Default): T|
     null {
   if (_currentInjector === undefined) {
-    throw new Error(`inject() must be called from an injection context`);
+    const errorMessage = typeof ngDevMode === 'undefined' || ngDevMode ?
+        `inject() must be called from an injection context` :
+        '';
+    throw new RuntimeError(RuntimeErrorCode.OUTSIDE_INJECTION_CONTEXT, errorMessage);
```
Comment (AndrewKushnir) [inline packages/core/src/di/injector_compatibility.ts:65]:
```suggestion
    throw new RuntimeError(RuntimeErrorCode.MISSING_INJECTION_CONTEXT, errorMessage);
```

File: packages/core/src/di/injector_compatibility.ts [hunk 8]
```diff
@@ -98,15 +102,15 @@ export function ɵɵinject<T>(token: ProviderToken<T>, flags = InjectFlags.Defau
...
       `This constructor is not compatible with Angular Dependency Injection because its dependency at index ${
           index} of the parameter list is invalid.
 This can happen if the dependency type is a primitive like a string or if an ancestor of this class is missing an Angular decorator.
 
 Please check that 1) the type for the parameter at index ${
           index} is correct and 2) the correct Angular decorators are defined for this class and its ancestors.` :
-      'invalid';
-  throw new Error(msg);
+      '';
+  throw new RuntimeError(RuntimeErrorCode.INVALID_FACTORY_DEPENDENCY, errorMessage);
```
Comment (AndrewKushnir) [inline packages/core/src/di/injector_compatibility.ts:113]: Same as above, let's revert this change (keep the code as is).

File: packages/core/src/di/injector_compatibility.ts [hunk 9]
```diff
@@ -141,7 +145,10 @@ export function injectArgs(types: (ProviderToken<any>|any[])[]): any[] {
     const arg = resolveForwardRef(types[i]);
     if (Array.isArray(arg)) {
       if (arg.length === 0) {
-        throw new Error('Arguments array must have arguments.');
+        const errorMessage = (typeof ngDevMode === 'undefined' || ngDevMode) ?
+            'Arguments array must have arguments.' :
+            '';
+        throw new RuntimeError(RuntimeErrorCode.INVALID_ARGUMENTS, errorMessage);
```
Comment (AndrewKushnir) [inline packages/core/src/di/injector_compatibility.ts:151]:
```suggestion
        throw new RuntimeError(RuntimeErrorCode.INVALID_INJECTION_ARGUMENTS, errorMessage);
```

File: packages/core/src/render3/error_code.ts [hunk 10]
```diff
@@ -72,9 +87,9 @@ export function formatRuntimeError(code: RuntimeErrorCode, message: string): str
   // Some runtime errors are still thrown without `ngDevMode` (for example
   // `throwProviderNotFoundError`), so we add `ngDevMode` check here to avoid pulling
   // `RUNTIME_ERRORS_WITH_GUIDES` symbol into prod bundles.
-  // TODO: revisit all instances where `RuntimeError` is thrown and see if `ngDevMode` can be added
-  // there instead to tree-shake more devmode-only code (and eventually remove `ngDevMode` check
-  // from this code).
+  // TODO: revisit all instances where `RuntimeError` is thrown and see if `ngDevMode` can be
+  // added there instead to tree-shake more devmode-only code (and eventually remove `ngDevMode`
+  // check from this code).
```
Comment (AndrewKushnir) [inline packages/core/src/render3/error_code.ts:92]: This change looks unnecessary, could you please revert?

File: packages/core/src/render3/features/inherit_definition_feature.ts [hunk 11]
```diff
@@ -39,7 +40,10 @@ export function ɵɵInheritDefinitionFeature(definition: DirectiveDef<any>|Compo
       superDef = superType.ɵcmp || superType.ɵdir;
     } else {
       if (superType.ɵcmp) {
-        throw new Error('Directives cannot inherit Components');
+        const errorMessage = typeof ngDevMode === 'undefined' || ngDevMode ?
+            'Directives cannot inherit Components' :
+            '';
+        throw new RuntimeError(RuntimeErrorCode.DIRECTIVES_CANNOT_INHERIT_COMPONENTS, errorMessage);
```
Comment (AndrewKushnir) [inline packages/core/src/render3/features/inherit_definition_feature.ts:46]:
```suggestion
        throw new RuntimeError(RuntimeErrorCode.INVALID_INHERITANCE, errorMessage);
```

File: packages/core/src/render3/i18n/i18n_apply.ts [hunk 12]
```diff
@@ -198,7 +200,10 @@ export function applyMutableOpCodes(
               attrValue, null);
           break;
         default:
-          throw new Error(`Unable to determine the type of mutate operation for "${opCode}"`);
+          const errorMessage = typeof ngDevMode === 'undefined' || ngDevMode ?
+              `Unable to determine the type of mutate operation for "${opCode}"` :
+              '';
+          throw new RuntimeError(RuntimeErrorCode.MUTABLE_OPCODES, errorMessage);
```
Comment (AndrewKushnir) [inline packages/core/src/render3/i18n/i18n_apply.ts:206]: In this code we can call `ngDevMode` without extra checks (see some usages above). We could also wrap the entire throw into `ngDevMode` check (it's ok to tree-shake away this in prod bundles):
```suggestion
          if (ngDevMode) {
            throw new RuntimeError(
              RuntimeErrorCode.INVALID_I18N_STRUCTURE,
              `Unable to determine the type of mutate operation for "${opCode}"`);
          }
```

File: packages/core/src/render3/view_ref.ts [hunk 13]
```diff
@@ -284,7 +286,10 @@ export class ViewRef<T> implements viewEngine_EmbeddedViewRef<T>, viewEngine_Int
 
   attachToViewContainerRef() {
     if (this._appRef) {
-      throw new Error('This view is already attached directly to the ApplicationRef!');
+      const errorMessage = typeof ngDevMode === 'undefined' || ngDevMode ?
+          'This view is already attached directly to the ApplicationRef!' :
+          '';
+      throw new RuntimeError(RuntimeErrorCode.VIEW_ALREADY_ATTACHED, errorMessage);
```
Comment (AndrewKushnir) [inline packages/core/src/render3/view_ref.ts:292]: It should be safe to call `ngDevMode` without extra checks here as well:
```suggestion
      const errorMessage = ngDevMode ?
          'This view is already attached directly to the ApplicationRef!' :
          '';
      throw new RuntimeError(RuntimeErrorCode.VIEW_ALREADY_ATTACHED, errorMessage);
```

File: packages/core/src/render3/view_ref.ts [hunk 14]
```diff
@@ -296,7 +301,10 @@ export class ViewRef<T> implements viewEngine_EmbeddedViewRef<T>, viewEngine_Int
 
   attachToAppRef(appRef: ViewRefTracker) {
     if (this._attachedToViewContainer) {
-      throw new Error('This view is already attached to a ViewContainer!');
+      const errorMessage = typeof ngDevMode === 'undefined' || ngDevMode ?
+          'This view is already attached to a ViewContainer!' :
+          '';
+      throw new RuntimeError(RuntimeErrorCode.VIEW_ALREADY_ATTACHED, errorMessage);
```
Comment (AndrewKushnir) [inline packages/core/src/render3/view_ref.ts:307]:
```suggestion
      const errorMessage = ngDevMode ?
          'This view is already attached to a ViewContainer!' :
          '';
      throw new RuntimeError(RuntimeErrorCode.VIEW_ALREADY_ATTACHED, errorMessage);
```

File: packages/core/src/sanitization/sanitization.ts [hunk 15]
```diff
@@ -117,7 +118,10 @@ export function ɵɵsanitizeResourceUrl(unsafeResourceUrl: any): TrustedScriptUR
   if (allowSanitizationBypassAndThrow(unsafeResourceUrl, BypassType.ResourceUrl)) {
     return trustedScriptURLFromStringBypass(unwrapSafeValue(unsafeResourceUrl));
   }
-  throw new Error('unsafe value used in a resource URL context (see https://g.co/ng/security#xss)');
+  const errorMessage = typeof ngDevMode === 'undefined' || ngDevMode ?
```
Comment (AndrewKushnir) [inline packages/core/src/sanitization/sanitization.ts:121]: I'd propose keeping parenthesis around the condition here (and in other cases in this PR where parenthesis are missing) for better readability:
```suggestion
  const errorMessage = (typeof ngDevMode === 'undefined' || ngDevMode) ?
```

File: packages/core/src/sanitization/sanitization.ts [hunk 16]
```diff
@@ -165,7 +171,10 @@ export function ɵɵtrustConstantHtml(html: TemplateStringsArray): TrustedHTML|s
   // template literal has no interpolation if and only if the length of the
   // TemplateStringsArray is 1.
   if (ngDevMode && (!Array.isArray(html) || !Array.isArray(html.raw) || html.length !== 1)) {
-    throw new Error(`Unexpected interpolation in trusted HTML constant: ${html.join('?')}`);
+    const errorMessage = typeof ngDevMode === 'undefined' || ngDevMode ?
+        `Unexpected interpolation in trusted HTML constant: ${html.join('?')}` :
+        '';
+    throw new RuntimeError(RuntimeErrorCode.UNEXPECTED_INTERPOLATION, errorMessage);
```
Comment (AndrewKushnir) [inline packages/core/src/sanitization/sanitization.ts:177]: This code is already in the `ngDevMode` check, so you can drop the check and just use error message as is.

File: packages/core/src/sanitization/sanitization.ts [hunk 17]
```diff
@@ -191,7 +200,10 @@ export function ɵɵtrustConstantResourceUrl(url: TemplateStringsArray): Trusted
   // array. The associated template literal has no interpolation if and only if
   // the length of the TemplateStringsArray is 1.
   if (ngDevMode && (!Array.isArray(url) || !Array.isArray(url.raw) || url.length !== 1)) {
-    throw new Error(`Unexpected interpolation in trusted URL constant: ${url.join('?')}`);
+    const errorMessage = typeof ngDevMode === 'undefined' || ngDevMode ?
+        `Unexpected interpolation in trusted URL constant: ${url.join('?')}` :
+        '';
+    throw new RuntimeError(RuntimeErrorCode.UNEXPECTED_INTERPOLATION, errorMessage);
```
Comment (AndrewKushnir) [inline packages/core/src/sanitization/sanitization.ts:206]: Same here, we are in the `ngDevMode` block, so we can use the error message as is.

File: packages/core/src/sanitization/sanitization.ts [hunk 18]
```diff
@@ -234,19 +246,23 @@ export function ɵɵsanitizeUrlOrResourceUrl(unsafeUrl: any, tag: string, prop:
 
 export function validateAgainstEventProperties(name: string) {
   if (name.toLowerCase().startsWith('on')) {
-    const msg = `Binding to event property '${name}' is disallowed for security reasons, ` +
-        `please use (${name.slice(2)})=...` +
-        `\nIf '${name}' is a directive input, make sure the directive is imported by the` +
-        ` current module.`;
-    throw new Error(msg);
+    const errorMessage = typeof ngDevMode === 'undefined' || ngDevMode ?
```
Comment (AndrewKushnir) [inline packages/core/src/sanitization/sanitization.ts:249]: Both functions below are called from within the blocks that are guarded by `ngDevMode`, so these functions are tree-shaken away already, so we should only add the RuntimeError class here.

File: packages/core/src/render3/error_code.ts [hunk 19]
```diff
@@ -42,6 +47,16 @@ export const enum RuntimeErrorCode {
   // i18n Errors
 
   // Compilation Errors
+
+  // Other
+  INVALID_ARGUMENTS = '900',
+  NO_SUPPORTING_FACTORY = '901',
+  VIEW_ALREADY_ATTACHED = '902',
+  DIRECTIVES_CANNOT_INHERIT_COMPONENTS = '903',
+  MUTABLE_OPCODES = '904',
```
Comment (AndrewKushnir) [inline packages/core/src/render3/error_code.ts:56]: I think this one should go under the `i18n Errors` section.

File: packages/core/src/render3/error_code.ts (hunk 1 above)
Comment (AndrewKushnir) [reply packages/core/src/render3/error_code.ts:52]: I've left some comments and proposals on the error codes. Could you please apply the changes and I'll take a look at the final list to figure out if we can re-group the new ones? Thank you.

**Review by AndrewKushnir — COMMENTED** [review]
@ramthir thanks for creating this PR 👍 I've left some comments, could you please take a look when you get a chance?

Also, according to the CI, this change brings ~600 bytes improvement 🎉

File: packages/core/src/sanitization/sanitization.ts (hunk 18 above)
Comment (ramthir) [reply packages/core/src/sanitization/sanitization.ts:249]: Yeah, I'll do that. Also will keep a look on this in future changes. 👍

File: packages/core/src/render3/error_code.ts [hunk 20]
```diff
@@ -40,8 +45,18 @@ export const enum RuntimeErrorCode {
...
   // Compilation Errors
+
+  // Other
+  INVALID_DIFFER_INPUT = '900',
+  NO_SUPPORTING_DIFFER_FACTORY = '901',
+  VIEW_ALREADY_ATTACHED = '902',
+  INVALID_INHERITANCE = '903',
+  UNSAFE_VALUE_IN_RESOURCE_URL = '904',
+  UNSAFE_VALUE_IN_SCRIPT = '905',
+  UNEXPECTED_INTERPOLATION = '906'
```
Comment (AndrewKushnir) [inline packages/core/src/render3/error_code.ts:59]: It looks like this code is unused.
```suggestion
```

**Review by AndrewKushnir — COMMENTED** [review]
@ramthir thanks for addressing the feedback, the grouping for the error codes look good (there is just 1 comment, see below). Could you please change the status of the PR from Draft -> Ready for Review, so that we complete the review? Thank you.

**Comment by ramthir** [comment]
@AndrewKushnir, I've made the requested changes. Would you please let me know if I need to make any other changes to this PR?

**Comment by AndrewKushnir** [comment]
@ramthir thanks for the updates. The changes look good and I'll have another look once PR #44348 lands (this PR would require some minor changes after that). I will keep you updated. Thank you.

**Comment by AndrewKushnir** [comment]
@ramthir just a quick update: we are in the process of discussing an update to the error code format to allow 5 digit codes (currently only 4 digit codes are supported) in a context of https://github.com/angular/angular/pull/44398. It would probably take some time, but we'd need to resolve that before we can proceed with this PR. There are no action items required at this moment, I just wanted to let you know that we'll get back to this PR once https://github.com/angular/angular/pull/44398 is resolved. Thank you.

**Comment by ramthir** [comment]
Thanks for taking the time to update @AndrewKushnir. Looking forward to this change, we can proceed in other areas. 

In the meantime, please let me know if there are any other areas we can improve on while waiting on this.

**Review by josephperrott — APPROVED** [review]
LGTM

**Comment by AndrewKushnir** [comment]
@ramthir just wanted to let you know that PR https://github.com/angular/angular/pull/44398 is now merged into the main branch and this PR is no longer blocked. The list of error code has moved to `packages/core/src/errors.ts` and error codes are now integers (vs strings). Could you please rebase this PR and update the code according to the current state? Please let us know if you have any questions. Thank you.

**Comment by ramthir** [comment]
@AndrewKushnir I have a test failure due to golden with `stringify6`; how do we proceed?

**Comment by AndrewKushnir** [comment]
> I have a test failure due to golden with stringify6; how do we proceed?

@ramthir please run the following command locally - it'd update a golden file as needed:

```
yarn bazel run //packages/core/test/bundling/router:symbol_test.accept
```

**Review by AndrewKushnir — APPROVED** [review]
@ramthir thanks for additional updates! The changes look good 👍

**Review by AndrewKushnir — APPROVED** [review]
Reviewed-for: public-api, size-tracking

**Comment by AndrewKushnir** [comment]
[Presubmit](http://test/OCL:421917265:BASE:421917303:1642200444894:e07643ad).

**Review by atscott — APPROVED** [review]
reviewed-for: public-api
reviewed-for: size-tracking

**Review by dylhunn — APPROVED** [review]
reviewed-for: public-api

**Review by alxhub — APPROVED** [review]
Reviewed-for: global-approvers

Two questions:

1) is it intentional that some error messages are guarded with a plain read of `ngDevMode` and others use a `typeof` guard before reading `ngDevMode`?
2) what do you think about altering the constructor of `RuntimeError` to allow for the `false` from `ngDevMode && 'error message'` to be passed through directly? That would eliminate the need for a temporary variable for the error string, as well as the need to use `''` explicitly in those cases.

**Comment by AndrewKushnir** [comment]
@alxhub thanks for the review 👍 

> 1. is it intentional that some error messages are guarded with a plain read of ngDevMode and others use a typeof guard before reading ngDevMode?

This is a ViewEngine artifact, we should be able to get rid of it, but it'd need some investigation (to see if this would be a breaking change for people that build their apps by custom tooling (outside of our CLI)).

> 2. what do you think about altering the constructor of RuntimeError to allow for the false from ngDevMode && 'error message' to be passed through directly? That would eliminate the need for a temporary variable for the error string, as well as the need to use '' explicitly in those cases.

That's an awesome idea 👍 I will update the `RuntimeError` class to support that and we can do more refactoring once we confirm that it's safe to use the `ngDevMode` flag without extra `typeof` checks.

**Comment by AndrewKushnir** [comment]
@ramthir I think we can proceed with the PR in its current state and I'll let you know once there are remaining pieces in place to do further refactoring as Alex proposed.

**Comment by AndrewKushnir** [comment]
@ramthir I've rebased this PR (there was 1 minor merge conflict) and this PR is now in the merge queue!

I wanted to say thanks for your help to standardize the error messages as well as making them more tree-shakable 👍

**Comment by dylhunn** [comment]
This PR was merged into the repository by commit bc03a2e3e3add26a0c52e1ae6eba8ca5f5da5786.
